### DISCUSSION TITLE
Option for postUpgradeTasks to run in packageFile directory

### DISCUSSION THREAD

**Discussion by cmplank** [renovatebot/renovate#13655] [Suggest an Idea]
My repo has subprojects in subdirectories with a single master build in the root. But most of the files that Renovate looks at are in subdirectories. Like this:

```
my-proj
    ui-app1
        - package.json
    ui-app2
        - package.json
    rest-app
        - build.gradle
    - build.gradle
```

I've had to do an obnoxious workaround to get my postUpgradeTasks config to run in the directory of the packageFile that they are being triggered for. My config looks like this:

```
    {
      matchPackagePrefix: ["@angular"],
      groupName: "angular monorepos together",
      postUpgradeTasks: {
        commands: [
          // Because my package.json is in a subproject directory, I have to cd to it prior to running commands.
          // The parentheses groups the commands so that the directory isn't changed in subsequent commands after the ending parens.
          "(cd {{{packageFileDir}}} && npm install --ignore-scripts --force)",
          "(cd {{{packageFileDir}}} && npx ng update {{{depName}}} --migrate-only --from={{{currentVersion}}} --to={{{newVersion}}} --allow-dirty --force)"
        ],
        fileFilters: ["**/**"]
      },
      separateMultipleMajor: true
    },
```

I propose creating a config option to tell the `postUpgradeTask` to run in the subdirectory with the `package.json` that it is triggered off of. It seems like it would be a common need and was pretty burdensome to update the shell commands, especially when considering that I had to update the whitelist as well. And the longer commands are harder to read and work with. And I now have duplicate entries in the whitelist (those with the `(cd \\S ` prefix and those without).

And thanks for all of your effort working on Renovate! I love this tool!

**Comment by rarkins** [comment]
Maybe a new option `postUpgradeTasks.cwd` which defaults to `""` but can be set to e.g. `{{{packageFileDir}}}`?

**Reply by cmplank** [reply]
That makes sense to me. I wouldn't have expected templating to be allowed in a field like that, but I see why it would be important if other values are just directory names.

**Reply by rarkins** [reply]
I thought of that approach because it's possible someone might have a `tools/` directory or something like that

**Reply by chkpnt** [reply]
First I thought using `npm --prefix ...` would work. But then an other npm version is used than described in `package.json`. And other tools might not have a prefix or working directory option at all. So yes: I endorse having an option `postUpgradeTasks.cwd`.

**Reply by dominik-weber** [reply]
We're also doing it similar to @cmplank's solution (I think the parentheses can be removed as the working directory does not seem to be preserved between commands anyway).
This works but is a bit clunky - especially because we also have to declare it in `allowedPostUpgradeCommands`.

+1 for `postUpgradeTasks.cwd` - It seems to me like this is an option that many people using multi project repos will need

**Comment by rarkins** [comment]
Converted to issue: https://github.com/renovatebot/renovate/issues/33124
